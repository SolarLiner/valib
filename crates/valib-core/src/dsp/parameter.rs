//! Shared values for passing parameters into DSP code.
//!
//! By design, those parameters are not used by the implementations in this
//! library. This is because it would be too limiting and cumbersome to go from
//! single-valued parameter values to potentially multi-valied values (eg. a
//! filter cutoff of a stereo filter, where the stereo signal is represented as
//! a single f32x2 SIMD type). They are provided as is for bigger modules, and
//! the traits implemented by "container" modules (like [`Series`] or
//! [`Parallel`]) to ease their use in propagating parameters.
//!
//! [`Series`]: crate::dsp::blocks::Series
//! [`Parallel`]: crate::dsp::blocks::Parallel
use std::borrow::Cow;
use std::marker::PhantomData;
use std::ops;
use std::sync::atomic::Ordering;
use std::sync::Arc;

use portable_atomic::{AtomicBool, AtomicF32};

pub use valib_derive::ParamName;

use crate::dsp::buffer::{AudioBufferMut, AudioBufferRef};
use crate::dsp::{DSPMeta, DSPProcess, DSPProcessBlock};
use crate::Scalar;

/// Filtered parameter value, useful with any DSP<1, 1, Sample=f32> algorithm.
pub struct FilteredParam<P> {
    /// Raw parameter value. This can be set directly to update the parameter.
    pub param: f32,
    /// Process which will take in the raw value and filter it.
    pub dsp: P,
}

impl<P: DSPProcess<1, 1, Sample = f32>> DSPMeta for FilteredParam<P> {
    type Sample = f32;
}

impl<P: DSPProcess<1, 1, Sample = f32>> FilteredParam<P> {
    /// Process the next sample generated from the parameter value.
    pub fn next_sample(&mut self) -> f32 {
        self.process([])[0]
    }
}

impl<P: DSPProcess<1, 1, Sample = f32>> DSPProcess<0, 1> for FilteredParam<P> {
    fn process(&mut self, _x: [Self::Sample; 0]) -> [Self::Sample; 1] {
        self.dsp.process([self.param])
    }
}

#[derive(Debug, Copy, Clone)]
enum Smoothing {
    Exponential {
        state: f32,
        fc: f32,
        lambda: f32,
    },
    Linear {
        samplerate: f32,
        last_out: f32,
        max_per_sec: f32,
    },
}

impl Smoothing {
    fn set_samplerate(&mut self, new_sr: f32) {
        match self {
            Self::Exponential { fc, lambda, .. } => {
                *lambda = *fc / new_sr;
            }
            Self::Linear { samplerate, .. } => {
                *samplerate = new_sr;
            }
        }
    }

    fn is_changing(&self, value: f32) -> bool {
        match self {
            Self::Exponential { state, .. } => (value - state).abs() < 1e-6,
            Self::Linear { last_out, .. } => (value - last_out).abs() < 1e-6,
        }
    }
}

impl DSPMeta for Smoothing {
    type Sample = f32;
}

impl DSPProcess<1, 1> for Smoothing {
    #[inline]
    fn process(&mut self, x: [Self::Sample; 1]) -> [Self::Sample; 1] {
        match self {
            Self::Exponential { state, lambda, .. } => {
                *state += (x[0] - *state) * *lambda;
                [*state]
            }
            Self::Linear {
                samplerate,
                last_out,
                max_per_sec,
            } => {
                let max_diff = *max_per_sec / *samplerate;
                let diff = x[0] - *last_out;
                *last_out += diff.clamp(-max_diff, max_diff);
                [*last_out]
            }
        }
    }
}

/// Smoothed parameter. Smoothing can be applied exponentially or linearly.
#[derive(Debug, Copy, Clone)]
pub struct SmoothedParam {
    /// Raw parameter value; can be set directly to change the target of the smoothed parameter.
    pub param: f32,
    smoothing: Smoothing,
}

impl DSPMeta for SmoothedParam {
    type Sample = f32;

    fn set_samplerate(&mut self, samplerate: f32) {
        self.smoothing.set_samplerate(samplerate);
    }
}

impl DSPProcess<0, 1> for SmoothedParam {
    #[inline]
    fn process(&mut self, _x: [Self::Sample; 0]) -> [Self::Sample; 1] {
        self.smoothing.process([self.param])
    }
}

impl SmoothedParam {
    /// Create a new linear smoothed parameter, with the given parameter and the maximum duration in milliseconds.
    ///
    /// # Arguments
    ///
    /// * `param`: Inner parameter to tap values from.
    /// * `samplerate`: Samplerate at which the smoother will run.
    /// * `duration_ms`: Maximum duration of a sweep, that is the duration it would take to go from one extreme to the other.
    pub fn linear(initial_value: f32, samplerate: f32, duration_ms: f32) -> Self {
        Self {
            param: initial_value,
            smoothing: Smoothing::Linear {
                samplerate,
                max_per_sec: duration_ms.recip(),
                last_out: initial_value,
            },
        }
    }

    /// Create a new exponential smoothed parameter, with the given parameter and the T60 time constant in milliseconds.
    ///
    /// # Arguments
    ///
    /// * `param`: Inner parameter to tap values from
    /// * `samplerate`: Samplerate parameter
    /// * `t60_ms`: "Time to decay by 60 dB" -- the time it takes for the output to be within 0.1% of the target value.
    pub fn exponential(initial_value: f32, samplerate: f32, t60_ms: f32) -> Self {
        let tau = 6.91 / t60_ms * 1e3;
        Self {
            param: initial_value,
            smoothing: Smoothing::Exponential {
                state: initial_value,
                fc: tau,
                lambda: tau / samplerate,
            },
        }
    }

    /// Returns the current smoothed value of the parameter.
    pub fn current_value(&self) -> f32 {
        match self.smoothing {
            Smoothing::Linear { last_out, .. } => last_out,
            Smoothing::Exponential { state, .. } => state,
        }
    }

    /// Computes the next sample of the smoother.
    pub fn next_sample(&mut self) -> f32 {
        self.process([])[0]
    }

    /// Computes the next sample of the smoother, casting it into a `T`.
    pub fn next_sample_as<T: Scalar>(&mut self) -> T {
        T::from_f64(self.next_sample() as _)
    }

    /// Returns true when the smoother is still in the process of smoothing the change to the raw value.
    pub fn is_changing(&self) -> bool {
        self.smoothing.is_changing(self.param)
    }
}

/// Parameter ID alias. Useful for type-erasing parameter names and make communication easier, but
/// this risks unwanted transmutations if not handled properly.
///
/// Note that transmuting between param ids is safe because both sides allow converting to and from
/// a [`ParamId`].
pub type ParamId = usize;

/// Trait for types that are parameter names.
///
/// This trait is most easily implemented as an enum of all possible parameters, but it allows
/// constructs such as `[P; N]` where `P: ParamName` and `const N: usize` to be defined for
/// automatic duplication of parameters, or 1->N communication of parameter values.
pub trait ParamName: Copy {
    /// Total number of elements in this type
    fn count() -> usize;

    /// Construct a [`Self`] from a [`ParamId`] value. The caller is expected to verify `value <
    /// Self::count()`, and so this method is declared as infallible.
    fn from_id(value: ParamId) -> Self;

    /// Construct a [`ParamId`] from this [`Self`].
    ///
    /// It is expected that round-trip conversion returns the same enum as the one we started with,
    /// that is:
    ///
    /// ```rust,compile_fail
    /// assert_eq!(self, Self::from_id(self.into_id()));
    /// ```
    ///
    /// where `Self: PartialEq`.
    fn into_id(self) -> ParamId;

    /// Return a user-friendly name for this parameter name.
    fn name(&self) -> Cow<'static, str>;

    /// Create an iterator returning all values for this type, that is, all values converted from
    /// IDs in sequence in the range `0..Self::count()`.
    fn iter() -> impl Iterator<Item = Self> {
        (0..Self::count()).map(|i| Self::from_id(i as _))
    }
}

/// Trait of types which have modulatable parameters.
pub trait HasParameters {
    /// Parameter name type
    type Name: Copy + ParamName;

    /// Set a new value for the parameter at the given parameter name.
    fn set_parameter(&mut self, param: Self::Name, value: f32);
}

/// Extension trait for types which have parameters.
pub trait HasParametersExt: HasParameters {
    /// Set the parameter as a boolean value. It will be encoded such that `value > 0.5` decodes
    /// back to the input boolean value.
    ///
    /// # Arguments
    ///
    /// * `param`: Name of the parameter to change
    /// * `value`: Boolean value to set
    ///
    /// returns: ()
    fn set_parameter_bool(&mut self, param: Self::Name, value: bool) {
        self.set_parameter(param, if value { 1.0 } else { 0.0 });
    }
}

impl<'a, P: HasParameters> HasParameters for &'a mut P {
    type Name = P::Name;

    fn set_parameter(&mut self, param: Self::Name, value: f32) {
        HasParameters::set_parameter(*self, param, value);
    }
}

impl<P: HasParameters> HasParameters for Box<P> {
    type Name = P::Name;

    fn set_parameter(&mut self, param: Self::Name, value: f32) {
        P::set_parameter(&mut *self, param, value);
    }
}

/// Dynamic parameter type which advertises as having `N` possible names.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Dynamic<const N: ParamId>(ParamId);

impl<const N: ParamId> Dynamic<N> {
    /// Create a new `Dynamic<N>` parameter name, checking the passed in ID for validity.
    ///
    /// # Arguments
    ///
    /// * `value`: Raw ID to use for this parameter type. The parameter is valid when `value < N`.
    ///
    /// returns: Option<Dynamic<{ N }>>
    pub fn new(value: ParamId) -> Option<Self> {
        (value < N).then_some(Self(value))
    }
}

impl<const N: ParamId> ParamName for Dynamic<N> {
    fn count() -> usize {
        N
    }

    fn from_id(value: ParamId) -> Self {
        Self::new(value).unwrap()
    }

    fn into_id(self) -> ParamId {
        self.0
    }

    fn name(&self) -> Cow<'static, str> {
        Cow::Owned(self.0.to_string())
    }
}

/// Specialized map type for storing values associated to parameters.
#[derive(Debug, Clone)]
pub struct ParamMap<P, T> {
    data: Vec<T>,
    __param: PhantomData<P>,
}

impl<P: ParamName, T: Default> Default for ParamMap<P, T> {
    fn default() -> Self {
        Self::new(|_| T::default())
    }
}

/// Type which implements [`Iterator`] listing the parameters and their associated value.
pub struct ParamMapIntoIter<P, T> {
    data: std::vec::IntoIter<T>,
    item: u64,
    __param: PhantomData<P>,
}

impl<P: ParamName, T> Iterator for ParamMapIntoIter<P, T> {
    type Item = (P, T);

    fn next(&mut self) -> Option<Self::Item> {
        let value = self.data.next()?;
        let i = self.item;
        self.item += 1;
        Some((P::from_id(i as _), value))
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        let len = P::count();
        (len, Some(len))
    }
}

impl<P: ParamName, T> ExactSizeIterator for ParamMapIntoIter<P, T> {
    fn len(&self) -> usize {
        P::count() as _
    }
}

impl<P: ParamName, T> IntoIterator for ParamMap<P, T> {
    type Item = (P, T);
    type IntoIter = ParamMapIntoIter<P, T>;

    fn into_iter(self) -> Self::IntoIter {
        ParamMapIntoIter {
            data: self.data.into_iter(),
            item: 0,
            __param: PhantomData,
        }
    }
}

impl<P: ParamName, T> ops::Index<P> for ParamMap<P, T> {
    type Output = T;

    fn index(&self, index: P) -> &Self::Output {
        &self.data[index.into_id()]
    }
}

impl<P: ParamName + Clone, T> ops::Index<&P> for ParamMap<P, T> {
    type Output = T;

    fn index(&self, index: &P) -> &Self::Output {
        &self.data[(*index).into_id()]
    }
}

impl<P: ParamName, T> ops::IndexMut<P> for ParamMap<P, T> {
    fn index_mut(&mut self, index: P) -> &mut Self::Output {
        &mut self.data[index.into_id()]
    }
}

impl<P: ParamName + Clone, T> ops::IndexMut<&P> for ParamMap<P, T> {
    fn index_mut(&mut self, index: &P) -> &mut Self::Output {
        &mut self.data[(*index).into_id()]
    }
}

impl<P: ParamName, T> ParamMap<P, T> {
    /// Create a new parameter map, filled in by the provided closure.
    ///
    /// # Arguments
    ///
    /// * `fill_fn`: Closure which is called for each parameter, and returns the associated value.
    ///
    /// returns: ParamMap<P, T>
    pub fn new(fill_fn: impl FnMut(P) -> T) -> Self {
        Self {
            data: Vec::from_iter(P::iter().map(fill_fn)),
            __param: PhantomData,
        }
    }

    /// Iterate over parameters and references to their values.
    pub fn iter(&self) -> impl '_ + Iterator<Item = (P, &T)> {
        self.data
            .iter()
            .enumerate()
            .map(|(i, x)| (P::from_id(i as _), x))
    }

    /// Iterate over parameters and mutable references to their values.
    pub fn iter_mut(&mut self) -> impl '_ + Iterator<Item = (P, &mut T)> {
        self.data
            .iter_mut()
            .enumerate()
            .map(|(i, x)| (P::from_id(i as _), x))
    }
}

/// Object-safe trait for types with parameters.
pub trait HasParametersErased {
    /// Set a parameter value by its raw param ID.
    ///
    /// # Arguments
    ///
    /// * `param_id`: Raw param ID to set.
    /// * `value`: Value to set the param ID with.
    ///
    /// returns: ()
    fn set_parameter_raw(&mut self, param_id: ParamId, value: f32);
}

/// Proxy parameter updates to another type. This allows thread-safe control of processors via their
/// parameters.
pub struct ParamsProxy<P: ParamName> {
    params: ParamMap<P, Arc<AtomicF32>>,
    param_changed: ParamMap<P, Arc<AtomicBool>>,
}

/// Type alias for the type that allows remote control of processors via their parameters.
pub type RemoteControl<P> = Arc<ParamsProxy<P>>;

impl<P: ParamName> ParamsProxy<P> {
    /// Create a new param proxy.
    pub fn new() -> Arc<Self> {
        let params = ParamMap::new(|_| Arc::new(AtomicF32::new(0.0)));
        let param_changed = ParamMap::new(|_| Arc::new(AtomicBool::new(false)));
        Arc::new(Self {
            params,
            param_changed,
        })
    }

    /// Set a parameter for a remote type.
    ///
    /// # Arguments
    ///
    /// * `param`: Parameter to set
    /// * `value`: Value to set
    ///
    /// returns: ()
    pub fn set_parameter(&self, param: P, value: f32) {
        self.param_changed[param].store(true, Ordering::SeqCst);
        self.params[param].store(value, Ordering::SeqCst);
    }

    fn get_update(&self, param: P) -> Option<f32> {
        let has_changed = self.param_changed[param]
            .compare_exchange(true, false, Ordering::SeqCst, Ordering::SeqCst)
            .unwrap_or(false);
        if has_changed {
            return Some(self.params[param].load(Ordering::SeqCst));
        }
        None
    }
}

/// Type which remote controls the type `P` through its [`RemoteControlled::proxy`].
pub struct RemoteControlled<P: HasParameters> {
    /// Remote-controlled type
    pub inner: P,
    /// Remote control proxy, which you can clone and send to another thread.
    pub proxy: RemoteControl<P::Name>,
    update_params_phase: f32,
    update_params_step: f32,
}

impl<P: HasParameters + DSPMeta> DSPMeta for RemoteControlled<P> {
    type Sample = P::Sample;
}

impl<P: HasParameters + DSPProcess<I, O>, const I: usize, const O: usize> DSPProcess<I, O>
    for RemoteControlled<P>
{
    fn process(&mut self, x: [Self::Sample; I]) -> [Self::Sample; O] {
        self.update_params_phase += self.update_params_step;
        if self.update_params_phase > 1.0 {
            self.update_params_phase -= 1.0;
            self.update_parameters();
        }

        self.inner.process(x)
    }
}

#[profiling::all_functions]
impl<P: HasParameters + DSPProcessBlock<I, O>, const I: usize, const O: usize> DSPProcessBlock<I, O>
    for RemoteControlled<P>
{
    fn process_block(
        &mut self,
        inputs: AudioBufferRef<Self::Sample, I>,
        outputs: AudioBufferMut<Self::Sample, O>,
    ) {
        self.update_params_phase += self.update_params_step * inputs.samples() as f32;
        if self.update_params_phase > 1.0 {
            self.update_parameters();
            self.update_params_phase = self.update_params_phase.fract();
        }
        self.inner.process_block(inputs, outputs);
    }

    fn max_block_size(&self) -> Option<usize> {
        self.inner.max_block_size()
    }
}

impl<P: HasParameters> RemoteControlled<P> {
    /// Create a new remote, controlling the passed in processor.
    ///
    /// # Arguments
    ///
    /// * `samplerate`: Sample rate at which the processor and remote control will run
    /// * `update_frequency`: Frequency (in Hz) at which the remote control will check for updated
    ///     parameters, and transfer them to the inner processor.
    /// * `inner`: Inner processor, that is going to be controlled by this.
    ///
    /// returns: RemoteControlled<P>
    pub fn new(samplerate: f32, update_frequency: f32, inner: P) -> Self {
        Self {
            inner,
            proxy: ParamsProxy::new(),
            update_params_phase: 0.0,
            update_params_step: update_frequency * samplerate.recip(),
        }
    }
}

#[profiling::all_functions]
impl<P: HasParameters> RemoteControlled<P> {
    /// Check for update on all parameters, and transmit them to the inner processor if they have
    /// changed.
    pub fn update_parameters(&mut self) {
        for param in P::Name::iter() {
            if let Some(value) = self.proxy.get_update(param) {
                self.inner.set_parameter(param, value);
            }
        }
    }
}
