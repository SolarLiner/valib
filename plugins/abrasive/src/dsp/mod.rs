use crate::{
    dsp::filter::{FilterModule, FilterParams},
    MAX_BLOCK_SIZE,
};
use nalgebra::Complex;
use nih_plug::prelude::*;
use std::sync::Arc;
use valib::Scalar;
use valib::{
    dsp::{analysis::DspAnalysis, blocks::Series, BlockAdapter},
    oversample::{Oversample, Oversampled},
};
use valib::{
    dsp::{DSPMeta, DSPProcess},
    simd::SimdComplexField,
};

pub mod filter;

pub const NUM_BANDS: usize = 5;

#[derive(Debug, Params)]
pub struct DspParams {
    #[id = "drive"]
    pub drive: FloatParam,
    #[id = "scale"]
    pub scale: FloatParam,
    #[nested(array)]
    pub filters: [Arc<FilterParams>; NUM_BANDS],
}

impl Default for DspParams {
    fn default() -> Self {
        Self {
            drive: FloatParam::new(
                "Drive",
                1.,
                FloatRange::Skewed {
                    min: 1e-2,
                    max: 100.,
                    factor: FloatRange::gain_skew_factor(-40., 40.),
                },
            )
            .with_string_to_value(formatters::s2v_f32_gain_to_db())
            .with_value_to_string(formatters::v2s_f32_gain_to_db(2))
            .with_unit("dB")
            .with_smoother(SmoothingStyle::Exponential(100.)),
            scale: FloatParam::new("Scale", 1., FloatRange::Linear { min: -1., max: 1. })
                .with_string_to_value(formatters::s2v_f32_percentage())
                .with_value_to_string(formatters::v2s_f32_percentage(2))
                .with_smoother(SmoothingStyle::Exponential(100.))
                .with_unit("%"),
            filters: std::array::from_fn(|_| Default::default()),
        }
    }
}

pub struct Equalizer<T> {
    params: Arc<DspParams>,
    dsp: Series<[FilterModule<T>; NUM_BANDS]>,
}

impl<T: Scalar> DSPMeta for Equalizer<T> {
    type Sample = T;
    fn set_samplerate(&mut self, samplerate: f32) {
        self.dsp.set_samplerate(samplerate);
    }

    fn latency(&self) -> usize {
        self.dsp.latency()
    }

    fn reset(&mut self) {
        self.dsp.reset()
    }
}

impl<T: Scalar> DSPProcess<1, 1> for Equalizer<T> {
    fn process(&mut self, [x]: [Self::Sample; 1]) -> [Self::Sample; 1] {
        let drive = T::from_f64(self.params.drive.smoothed.next() as _);
        let scale = T::from_f64(self.params.scale.smoothed.next() as _);
        for filter in &mut self.dsp.0 {
            filter.set_scale(scale);
        }
        let [y] = self.dsp.process([drive * x]);
        [y / drive]
    }
}

impl<T: Scalar> DspAnalysis<1, 1> for Equalizer<T> {
    fn h_z(&self, z: Complex<Self::Sample>) -> [[Complex<Self::Sample>; 1]; 1] {
        self.dsp.h_z(z)
    }

    fn freq_response(
        &self,
        samplerate: Self::Sample,
        f: Self::Sample,
    ) -> [[Complex<Self::Sample>; 1]; 1]
    where
        Complex<Self::Sample>: SimdComplexField,
    {
        self.dsp.freq_response(samplerate, f)
    }
}

impl<T: Scalar> Equalizer<T> {
    pub fn new(samplerate: T, params: Arc<DspParams>) -> Self {
        let dsp = Series(std::array::from_fn(|i| {
            FilterModule::new(samplerate, params.filters[i].clone())
        }));
        Self { params, dsp }
    }

    pub fn use_param_values(&mut self, use_modulated: bool) {
        for filter in &mut self.dsp.0 {
            filter.use_param_values(use_modulated);
        }
    }
}

pub const OVERSAMPLE: usize = 4;

pub type Dsp<T> = Oversampled<T, BlockAdapter<Equalizer<T>>>;
pub fn create<T: Scalar>(samplerate: f32, params: Arc<DspParams>) -> Dsp<T>
where
    Complex<T>: SimdComplexField,
{
    let target_samplerate = OVERSAMPLE as f32 * samplerate;
    Oversample::new(OVERSAMPLE, MAX_BLOCK_SIZE).with_dsp(
        samplerate,
        BlockAdapter(Equalizer::new(T::from_f64(target_samplerate as _), params)),
    )
}
