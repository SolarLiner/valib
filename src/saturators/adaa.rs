use num_traits::Zero;
use numeric_literals::replace_float_literals;
use simba::simd::SimdBool;

use crate::dsp::DSP;
use crate::saturators::{Asinh, Blend, Clipper, Saturator, Tanh};
use crate::Scalar;

/// Trait for functions that have antiderivatives.
pub trait Antiderivative<T> {
    fn evaluate(&self, x: T) -> T;

    fn antiderivative(&self, x: T) -> T;
}

pub trait Antiderivative2<T>: Antiderivative<T> {
    fn antiderivative2(&self, x: T) -> T;
}

impl<T: Scalar> Antiderivative<T> for Tanh {
    fn evaluate(&self, x: T) -> T {
        x.simd_tanh()
    }

    fn antiderivative(&self, x: T) -> T {
        x.simd_cosh().simd_ln()
    }
}

impl<T: Scalar> Antiderivative<T> for Asinh {
    fn evaluate(&self, x: T) -> T {
        x.simd_asinh()
    }

    fn antiderivative(&self, x: T) -> T {
        let x0 = x * x + T::one();
        x * x.simd_asinh() - x0.simd_sqrt()
    }
}

impl<T: Scalar> Antiderivative2<T> for Asinh {
    #[replace_float_literals(T::from_f64(literal))]
    fn antiderivative2(&self, x: T) -> T {
        let x0 = x * x + 1.0;
        let x1 = 2.0 * x * x - 1.0;
        let x2 = -x * 3.0 * x0.simd_sqrt() / 4.0;
        let x3 = x1 * x.simd_asinh() / 4.0;
        x2 + x3
    }
}

impl<T: Scalar> Antiderivative<T> for Clipper {
    #[replace_float_literals(T::from_f64(literal))]
    fn evaluate(&self, x: T) -> T {
        x.simd_clamp(-1.0, 1.0)
    }

    #[replace_float_literals(T::from_f64(literal))]
    fn antiderivative(&self, x: T) -> T {
        let lower = -x;
        let upper = x;
        let middle = (0.5) * (x * x + T::one());
        let is_lower = x.simd_lt(-T::one());
        let is_higher = x.simd_gt(T::one());
        lower.select(is_lower, upper.select(is_higher, middle))
    }
}

impl<T: Scalar> Antiderivative2<T> for Clipper {
    #[replace_float_literals(T::from_f64(literal))]
    fn antiderivative2(&self, x: T) -> T {
        let lower = -0.5 * x * (x - 2.0);
        let upper = x * x / 2.0 * x * T::simd_recip(3.0);
        let middle = T::simd_recip(6.0) * (x * x * x + 9.0 * x + 1.0);
        let is_lower = x.simd_lt(-1.0);
        let is_higher = x.simd_gt(1.0);
        lower.select(is_lower, upper.select(is_higher, middle))
    }
}

impl<T: Scalar, S: Antiderivative<T>> Antiderivative<T> for Blend<T, S> {
    fn evaluate(&self, x: T) -> T {
        x + (self.inner.evaluate(x) - x) * self.amt
    }

    #[replace_float_literals(T::from_f64(literal))]
    fn antiderivative(&self, x: T) -> T {
        let ad_x = 0.5 * x * x;
        self.inner.antiderivative(x) * self.amt + ad_x * (1.0 + self.amt)
    }
}

impl<T: Scalar, S: Antiderivative2<T>> Antiderivative2<T> for Blend<T, S> {
    #[replace_float_literals(T::from_f64(literal))]
    fn antiderivative2(&self, x: T) -> T {
        let ad_x = x + x * x * x / 6.0;
        self.inner.antiderivative2(x) * self.amt + ad_x * (1.0 + self.amt)
    }
}

#[derive(Debug, Copy, Clone)]
pub struct Adaa<T, S, const ORDER: usize> {
    pub epsilon: T,
    pub inner: S,
    memory: [T; ORDER],
}

impl<T: Scalar, S, const ORDER: usize> Adaa<T, S, ORDER> {
    pub fn new(inner: S) -> Self {
        Self {
            epsilon: T::from_f64(1e-3),
            inner,
            memory: [T::zero(); ORDER],
        }
    }
}

impl<T: Scalar, S: Default, const ORDER: usize> Default for Adaa<T, S, ORDER> {
    fn default() -> Self {
        Self::new(S::default())
    }
}

impl<T: Scalar, S: Antiderivative<T>> Adaa<T, S, 1> {
    #[replace_float_literals(T::from_f64(literal))]
    pub fn next_sample_immutable(&self, x: T) -> T {
        let den = x - self.memory[0];
        let below = den.simd_abs().simd_lt(self.epsilon);
        below.if_else(
            || self.inner.evaluate((x + self.memory[0]) / 2.0),
            || {
                let num = self.inner.antiderivative(x) - self.inner.antiderivative(self.memory[0]);
                num / den
            },
        )
    }

    pub fn commit_sample(&mut self, x: T) {
        self.memory = [x];
    }

    pub fn next_sample(&mut self, x: T) -> T {
        let y = self.next_sample_immutable(x);
        self.commit_sample(x);
        y
    }
}

impl<T: Scalar, S: Antiderivative<T> + Saturator<T>> Saturator<T> for Adaa<T, S, 1> {
    fn saturate(&self, x: T) -> T {
        self.next_sample_immutable(x)
    }

    fn update_state(&mut self, x: T, y: T) {
        self.commit_sample(x);
        self.inner.update_state(x, y);
    }

    fn sat_diff(&self, x: T) -> T {
        self.inner.sat_diff(x)
    }
}

impl<T: Scalar, S: Antiderivative<T> + DSP<1, 1, Sample = T>> DSP<1, 1> for Adaa<T, S, 1> {
    type Sample = T;

    fn process(&mut self, [x]: [Self::Sample; 1]) -> [Self::Sample; 1] {
        [self.next_sample(x)]
    }
}

impl<T: Scalar, S: Antiderivative2<T> + Saturator<T>> Saturator<T> for Adaa<T, S, 2> {
    #[replace_float_literals(T::from_f64(literal))]
    fn saturate(&self, x: T) -> T {
        let [x1, x2] = self.memory;
        let den1 = x - x1;
        let den2 = x1 - x2;
        let den3 = x - x2;
        let below1 = den1.simd_abs().simd_lt(self.epsilon);
        let below2 = den2.simd_abs().simd_lt(self.epsilon);
        let below3 = den3.simd_abs().simd_lt(self.epsilon);
        (below1 | below2 | below3).if_else(
            || self.inner.evaluate((x + x1) / 2.0),
            || {
                let num1 = self.inner.antiderivative(x) - self.inner.antiderivative2(x1);
                let num2 = self.inner.antiderivative2(x1) - self.inner.antiderivative2(x2);
                den3.simd_recip() * (num1 / den1 + num2 / den2)
            },
        )
    }

    fn update_state(&mut self, x: T, y: T) {
        self.inner.update_state(x, y);
        self.memory.swap(0, 1);
        self.memory[0] = x;
    }

    fn sat_diff(&self, x: T) -> T {
        self.inner.sat_diff(x)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;
    use std::f64::consts::TAU;

    #[rstest]
    #[case("tanh", Adaa::<_, Tanh, 1>::default())]
    #[case("asinh", Adaa::<_, Asinh, 1>::default())]
    #[case("clipper", Adaa::<_, Clipper, 1>::default())]
    fn test_adaa1<S: Antiderivative<f64> + Saturator<f64>>(
        #[case] name: &str,
        #[case] mut adaa: Adaa<f64, S, 1>,
    ) {
        let samplerate = 100.0;
        let f = 10.0;
        let input: [_; 100] =
            std::array::from_fn(|i| i as f64 / samplerate).map(|t| 3.0 * f64::sin(TAU * f * t));
        let output = input.map(|x| {
            let y = adaa.saturate(x);
            adaa.update_state(x, y);
            y
        });

        let name = format!("test_adaa1_{name}");
        insta::assert_csv_snapshot!(name, &output as &[_], { "[]" => insta::rounded_redaction(3) })
    }

    #[rstest]
    #[case("clipper", Adaa::<_, Clipper, 2>::default())]
    #[case("asinh", Adaa::<_, Asinh, 2>::default())]
    fn test_adaa2<S: Antiderivative2<f64> + Saturator<f64>>(
        #[case] name: &str,
        #[case] mut adaa: Adaa<f64, S, 2>,
    ) {
        let samplerate = 100.0;
        let f = 10.0;
        let input: [_; 100] =
            std::array::from_fn(|i| i as f64 / samplerate).map(|t| 3.0 * f64::sin(TAU * f * t));
        let output = input.map(|x| {
            let y = adaa.saturate(x);
            adaa.update_state(x, y);
            y
        });

        let name = format!("test_adaa2_{name}",);
        insta::assert_csv_snapshot!(name, &output as &[_], { "[]" => insta::rounded_redaction(3) })
    }
}
