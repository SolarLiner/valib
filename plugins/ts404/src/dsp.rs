use valib::dsp::DSP;
use valib::filters::statespace::StateSpace;
use valib::saturators::Slew;
use valib::Scalar;

#[derive(Debug, Copy, Clone)]
pub struct ClipperStage<T: Scalar>(StateSpace<T, 1, 3, 1>, Slew<T>);

impl<T: Scalar> DSP<1, 1> for ClipperStage<T> {
    type Sample = T;

    fn process(&mut self, x: [Self::Sample; 1]) -> [Self::Sample; 1] {
        let [y] = self.0.process(x);
        self.1.process([y.simd_asinh()])
    }

    fn latency(&self) -> usize {
        self.0.latency() + self.1.latency()
    }

    fn reset(&mut self) {
        self.0.reset();
        self.1.reset();
    }
}

impl<T: Scalar> ClipperStage<T> {
    pub fn new(samplerate: T, dist: T) -> Self {
        let dt = samplerate.simd_recip();
        Self(crate::gen::clipper(dist, dt), Slew::new(T::from_f64(1e5) * dt))
    }

    pub fn set_params(&mut self, samplerate: T, dist: T) {
        let dt = samplerate.simd_recip();
        self.0.update_matrices(&crate::gen::clipper(dist, dt));
        self.1.set_max_diff(T::from_f64(1e5), dt);
    }
}

#[derive(Debug, Copy, Clone)]
pub struct ToneStage<T: Scalar>(StateSpace<T, 1, 4, 1>);

impl<T: Scalar> DSP<1, 1> for ToneStage<T> {
    type Sample = T;

    fn process(&mut self, x: [Self::Sample; 1]) -> [Self::Sample; 1] {
        self.0.process(x)
    }

    fn latency(&self) -> usize {
        self.0.latency()
    }

    fn reset(&mut self) {
        self.0.reset();
    }
}

impl<T: Scalar> ToneStage<T> {
    pub fn new(samplerate: T, tone: T) -> Self {
        Self(crate::gen::tone(tone, samplerate.simd_recip()))
    }

    pub fn update_params(&mut self, samplerate: T, tone: T) {
        self.0.update_matrices(&crate::gen::tone(tone, samplerate.simd_recip()));
    }
}