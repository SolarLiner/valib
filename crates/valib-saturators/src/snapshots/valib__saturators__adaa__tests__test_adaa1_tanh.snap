---
source: src/saturators/adaa.rs
expression: "&output as &[_]"
---
0.0
0.623
0.976
0.993
0.976
0.623
-0.623
-0.976
-0.993
-0.976
-0.623
0.623
0.976
0.993
0.976
0.623
-0.623
-0.976
-0.993
-0.976
-0.623
0.623
0.976
0.993
0.976
0.623
-0.623
-0.976
-0.993
-0.976
-0.623
0.623
0.976
0.993
0.976
0.623
-0.623
-0.976
-0.993
-0.976
-0.623
0.623
0.976
0.993
0.976
0.623
-0.623
-0.976
-0.993
-0.976
-0.623
0.623
0.976
0.993
0.976
0.623
-0.623
-0.976
-0.993
-0.976
-0.623
0.623
0.976
0.993
0.976
0.623
-0.623
-0.976
-0.993
-0.976
-0.623
0.623
0.976
0.993
0.976
0.623
-0.623
-0.976
-0.993
-0.976
-0.623
0.623
0.976
0.993
0.976
0.623
-0.623
-0.976
-0.993
-0.976
-0.623
0.623
0.976
0.993
0.976
0.623
-0.623
-0.976
-0.993
-0.976
